Support for hot upgrades

guidotripaldi opened the pull request:
## Description

Adds hot upgrade support and documentation. Closes #195.

## Licensing/Copyright

**By submitting this PR, you agree to the following statement, please read before submission!**

I certify that I own, and have sufficient rights to contribute, all source code and
related material intended to be compiled or integrated with the source code for Bootleg
(the "Contribution"). My Contribution is licensed under the MIT License.

NOTE: If you submit a PR and remove the statement above, your PR will be rejected. For your PR to be
considered, it must contain your agreement to license under the MIT license.

guidotripaldi committed:
Adds (preliminary) hot upgrade support

guidotripaldi committed:
Removes added code

guidotripaldi committed:
Adds hot upgrade support to updated task/build/remote.exs

guidotripaldi committed:
Adds hot upgrade documentation

guidotripaldi committed:
Changes the heredoc identation

guidotripaldi committed:
Formats files

rjanja commented:
Hi @guidotripaldi, thanks for the PR! This looks great, and the documentation is fantastic!

This branch is now behind `master` and needs to be rebased in order to benefit from some testing pipeline improvements we've recently made. I'm happy to do so but this would require force-pushing to your branch, something you might not be expecting and that may cause a conflict for you.

After rebasing, the two missing pieces I see are
- Documentation (not the moduledocs, which look great, but a separate [docs](https://hexdocs.pm/bootleg/) section on hot upgrades)
- Functional tests

I'm happy to work on either or both of the above and help get this across the line. Thanks again!

guidotripaldi committed:
Fixs typos

guidotripaldi committed:
Adds hot_upgrade to MkDocs documentation

guidotripaldi commented:
Hello @rjanja , thank you!
I have rebased the `hot_upgrade` branch and forced a push to origin, now there should be no more conflicts.
I've added also the mkdocs documentation, please let me know if it fits well, or if you prefer a different style/structure.

rjanja commented:
Looks great @guidotripaldi ! Thanks for all your work on this. I'll spend some time with it over the weekend.

rjanja commented:
I apologize it's taking me longer to review this than I expected. I'll be pushing up some documentation changes soon but still need to work out the functional tests. On the bright side, I've been able to use this to do some hot upgrades of a really simple app :)

guidotripaldi commented:
no problem, take your time

lessless commented:
Hey guys,

Just don't want review of that amazing feature to fold over the Christmas ;)

lessless commented:
Hi @rjanja,  @guidotripaldi 

Would you please add a bit more gas to get this through review process? 

Thanks.

guidotripaldi commented:
Hello @rjanja, is there something I can do to help you in the review process for this long pending PR?

> Hi @rjanja, @guidotripaldi
> 
> Would you please add a bit more gas to get this through review process?
> 
> Thanks.

holetse commented:
@guidotripaldi Thanks for taking the time to pull this together, and for sticking with it for all these months. We'd like to incorporate these changes but there are still two outstanding issues:

- Conflicts with the master branch the must be resolved (please use a rebase) prior to merging.
- The new functionality needs unit and integration tests.

The tests are the bigger hurdle, as the conflicts look trivial. Unfortunately we do not have the time right now to add the missing test coverage in-house, but we do have the time to support someone else adding it. Examples on how to test tasks can be found here (both unit and functional tests): https://github.com/labzero/bootleg/tree/master/test/bootleg/tasks. Our goal is 100% coverage for these sorts of changes, as reported by coveralls.

lessless commented:
hey @guidotripaldi - let me know if you would like to pair on that

guidotripaldi commented:
> hey @guidotripaldi - let me know if you would like to pair on that

Hello lessless, Thank you for being available. Unfortunately I have no time at the moment to learn how to develop the required unit and integration tests, so it would be wonderful if you can take over this job

lessless commented:
Same thing here. I would love to use hot upgrades in bootleg but don't have time to figure out what exactly should be done here. 

@holetse I guess your situation didn't change as well, isn't it?

lessless commented:
@holetse @rjanja can you guys please help to define expectations and I'll try to find some time to get this merged in.

lessless commented:
@spinlock99  can you please share a reason to close this issue?

spinlock99 commented:
Hey @lessless. I just volunteered to try to clean up the bootleg repo and try to get things working again. Given that there hadn't been any updates on this in the last 5 years and the elixir community has moved away from hot upgrades, I thought it was dead. 

Are you still interested in supporting this? I've given up on Distillery myself but I'm happy to reopen the PR if you can do the heavy lifting.

lessless commented:
@spinlock99 gotcha. I'm currently not doing HCU, so can only cheer from the benches. 

I'd say that community never caught with HCU. There were some attempts by Chris McCord & Fly to get it off the ground but I don't think they saw a great success.  Which is a shame because it would be great to be able to have zero downtime deployments with LiveView. They still try though https://xcancel.com/yevkurtov/status/2029914829708750927
